chapters/grid-search.qmd

```diff
@@ -384,7 +383,497 @@ A simple workaround for creating a space-filling design is to repeat the unique
+
+::: {.important-box}
+It is important to emphasize that nested resampling is for **verification, not optimization**. It provides a better estimate of our model optimization process; it does not replace it. 
+:::
+
+We might best communicate the results of nested resampling like this: 
+
+> We tuned the learning rate of our boosted classification tree using grid search and found that a rate of 10<sup>`r signif(log10(grid_brier$learn_rate), 3)`</sup> to be best. We think that this model has a corresponding Brier score, measured without optimization bias, of `r signif(nest_grid_brier$mean, 3)`. 
+
+We can look at the candidates produced by the inner resamples to understand the stability of the optimization process and potentially diagnose other issues. We _would not_ choose "the best" inner resampling result and move forward with its candidate value. 
+
+It is a good idea to use nested resampling when the training set is small, the predictor set is very large, or both. The lack of significant bias in the analysis above does not discount the problem of optimization bias. It exists but can sometimes be within the experimental noise. 
+
+It is a good idea to choose an inner resampling scheme that is as close as possible to the single resampling scheme that you used for optimization. The outer scheme can vary depending on your needs. Care should be used when the outer resampling method is the bootstrap. Since it replicates training set points in its analysis sets, the inner resamples need to use the _unique_ rows of the original training set. Otherwise the same data point might end up in both the inner analysis and assessment sets. 
+
+The primary downside to nested resampling is the computational costs. Two layers of resampling have a quadratic cost, and only the inner resamples can be executed in parallel. Using `r num_workers` workers, the analysis in @fig-nested took `r sprintf("%2.1f", grid_nest_time[3]/grid_time[3])`-fold longer to compute than the basic grid search that produced @fig-1d-boost.
+
+## Setting Parameter Ranges
+
+Specifying the range of parameters use to define the grid may involve some guesswork. For some parameter types, there is a well understood and defined range. For others, such as learning rate, there is a lower bound (zero) and a loosely defined upper bound mostly based on convention and prior experience. For grids, we want to avoid configurations with few unique values and a wide range. This might not sample the section of the parameter space that includes nonlinearity and the region of optimal results. @fig-learn-rate-seq shows an example of this for the learning rate. The initial grid included three points that missed the regions of interest. @fig-learn-rate-grid did a better job with more grid points. In high dimensional parameter space, the likelihood of a poor grid increases, especially for relatively “small” grid sizes. 
```

Comment:
```suggestion
Specifying the range of parameters used to define the grid may involve some guesswork. For some parameter types, there is a well understood and defined range. For others, such as learning rate, there is a lower bound (zero) and a loosely defined upper bound mostly based on convention and prior experience. For grids, we want to avoid configurations with few unique values and a wide range. This might not sample the section of the parameter space that includes nonlinearity and the region of optimal results. @fig-learn-rate-seq shows an example of this for the learning rate. The initial grid included three points that missed the regions of interest. @fig-learn-rate-grid did a better job with more grid points. In high dimensional parameter space, the likelihood of a poor grid increases, especially for relatively “small” grid sizes. 
```